File: 16/day16.pl
:- ['../lib/io.pl'].

% assumes all input is positive
join_domains(Domains, AllowedDomain) :-
    foldl([X,Y,Z]>>(
        Z = Y\/X
    ), Domains, -1, AllowedDomain).

part1(Fields, Tickets, Ans) :-
    maplist([[_,D],X]>>(X=D), Fields, Domains),
    join_domains(Domains, AllowedDomain),
    maplist({AllowedDomain}/[List, Sum]>>(
        exclude([X]>>(X in AllowedDomain), List, Inv),
        sum(Inv, #=, Sum)
    ), Tickets, Invalid),
    sum(Invalid, #=, Ans).

part2(Fields, Ticket, Tickets, Ans) :-
    maplist([[_,D],X]>>(X=D), Fields, Domains),
    join_domains(Domains, AllowedDomain),
    include([X]>>(
        forall(member(M, X), (
            M in AllowedDomain
        ))
    ), Tickets, ValidTickets),

    length(Fields, N),
    length(Values, N),
    maplist(=([]), Values),
    foldl([X,Y,Z]>>(
        maplist([A,B,C]>>(C = [A|B]), X, Y, Z)
    ), [Ticket|ValidTickets], Values, AllValues),

    maplist({Fields}/[AV,List]>>(
        findall(M, (nth1(M, Fields, [_,FD]), AV ins FD), List)
    ), AllValues, AllValidFields),

    find_valid(AllValidFields, [], ValidConfig),
    numlist(1, 6, DepartureFields),
    maplist({ValidConfig, Ticket}/[X,Y]>>(
        member(I-X, ValidConfig),
        nth1(I, Ticket, Y)
    ), DepartureFields, TicketValues),
    foldl([X,Y,Z]>>(Z #= X * Y), TicketValues, 1, Ans).

find_valid(Configs, Found, Valid) :-
    findall(M-F, nth1(M, Configs, [F]), List),
    ( List = [] -> Valid = Found ; 
        append(List, Found, NewFound),
        maplist({NewFound}/[X,Y]>>(
            exclude([Z]>>(member(_-Z, NewFound)), X, Y)
        ), Configs, Filtered),
        find_valid(Filtered, NewFound, Valid)
    ).
    
parse(Fields, Ticket, Tickets) --> 
    parse_fields(Fields), "your ticket:\n", parse_ticket(Ticket),
    "\nnearby tickets:\n", parse_tickets(Tickets).

parse_fields([]) --> "\n".
parse_fields([[Str,R1\/R2]|T]) -->
    parse_field(Name, R1, R2), "\n", parse_fields(T), {string_codes(Str, Name)}.
parse_field(Fieldname, R1, R2) -->
    string_without(":", Fieldname), ": ", parse_range(R1), " or ", parse_range(R2).
parse_range(From..To) -->
    integer(From), "-", integer(To).

parse_ticket([]) --> "\n".
parse_ticket([H|T]) --> integer(H), parse_ticket(T).
parse_ticket([H|T]) --> ",", integer(H), parse_ticket(T).

parse_tickets([]) --> blanks, eos.
parse_tickets([H|T]) --> parse_ticket(H), parse_tickets(T).

run :-
    input_stream(16, parse(Fields, Ticket, Tickets)),
    part1(Fields, Tickets, Ans1),
    write_part1(Ans1),
    part2(Fields, Ticket, Tickets, Ans2),
    write_part2(Ans2).

File: 16/day16.go
package main

import (
    "fmt"
    "strconv"
    "strings"

    "github.com/deosjr/adventofcode2020/lib"
)

type ticket []int

type field struct {
    name string
    min1, max1, min2, max2 int
}

func (f field) inRange(value int) bool {
    return (f.min1 <= value && value <= f.max1) || (f.min2 <= value && value <= f.max2)
}

func (f field) allInRange(values []int) bool {
    for _, v := range values {
        if !f.inRange(v) {
            return false
        }
    }
    return true
}

func parseFields(str string) []field {
    fields := []field{}
    for _, line := range strings.Split(str, "\n") {
        split := strings.Split(line, ":")
        name := split[0]
        var min1, max1, min2, max2 int
        fmt.Sscanf(split[1], " %d-%d or %d-%d", &min1, &max1, &min2, &max2)
        fields = append(fields, field{name, min1, max1, min2, max2})
    }
    return fields
}

func parseTicket(str string) ticket {
    t := []int{}
    for _, s := range strings.Split(str, ",") {
        n, _ := strconv.Atoi(s)
        t = append(t, n)
    }
    return t
}

func parseTickets(str string) []ticket {
    tickets := []ticket{}
    for _, line := range strings.Split(str, "\n")[1:] {
        if len(line) == 0 {
            continue
        }
        tickets = append(tickets, parseTicket(line))
    }
    return tickets
}

func findInvalid(fields []field, tickets []ticket) (sum int) {
    for _, t := range tickets {
        n, _ := isTicketInvalid(fields, t)
        sum += n
    }
    return
}

func isTicketInvalid(fields []field, t ticket) (sum int, invalid bool) {
Loop:
    for _, v := range t {
        for _, f := range fields {
            if f.inRange(v) {
                continue Loop
            }
        }
        invalid = true
        sum += v
    }
    return
}

func findDepartureFields(fields []field, yt ticket, tickets []ticket) int {
    validTickets := []ticket{}
    for _, t := range tickets {
        if _, ok := isTicketInvalid(fields, t); ok {
            continue
        }
        validTickets = append(validTickets, t)
    }

    // gather all the known ticket values by column
    columns := make([][]int, len(yt))
    for i, v := range yt {
        columns[i] = []int{v}
    }
    for _, t := range validTickets {
        for i, v := range t {
            columns[i] = append(columns[i], v)
        }
    }

    // per column, find all fields for which all values are in range
    configs := []map[int]struct{}{}
    for _, values := range columns {
        validfields := map[int]struct{}{}
        for i, f := range fields {
            if f.allInRange(values) {
                validfields[i] = struct{}{}
            }
        }
        configs = append(configs, validfields)
    }

    // find all columns that can only fit one field, their mapping is now known
    // delete that field from the rest of configs and repeat until all are found
    m := map[int]int{}
    for len(m) < len(yt) {
        found := map[int]struct{}{}
        for i, cfg := range configs {
            if len(cfg) == 1 {
                for k,_ := range cfg {
                    m[k] = i
                    found[k] = struct{}{}
                }
            }
        }
        for _, cfg := range configs {
            for k, _ := range found {
                delete(cfg, k)
            }
        }
    }

    // departure fields are the first 6 fields in your ticket
    ans := 1
    for i:=0; i<6; i++ {
        ans *= yt[m[i]]
    }
    return ans
}

func main() {
    input := lib.ReadFile(16)
    split := strings.Split(input, "\n\n")
    fields := parseFields(split[0])
    ticketInput := strings.Split(split[1], "\n")[1]
    yourTicket := parseTicket(ticketInput)
    tickets := parseTickets(split[2])

    p1 := findInvalid(fields, tickets)
    lib.WritePart1("%d", p1)

    p2 := findDepartureFields(fields, yourTicket, tickets)
    lib.WritePart2("%d", p2)
}
